File: backend/src/main/java/com/safebite/backend/service/NoticiaService.java
package com.safebite.backend.service;

import com.safebite.backend.dto.request.NoticiaRequest;
import com.safebite.backend.dto.response.NoticiaResponse;
import com.safebite.backend.exception.ResourceNotFoundException;
import com.safebite.backend.model.Noticia;
import com.safebite.backend.model.Usuario;
import com.safebite.backend.repository.NoticiaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Comparator;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NoticiaService {

    private final NoticiaRepository noticiaRepository;

    public List<NoticiaResponse> listar(String etiqueta) {
        List<Noticia> noticias = noticiaRepository.findAll();
        return noticias.stream()
                .filter(n -> etiqueta == null || etiqueta.isBlank()
                        || n.getEtiquetas().stream().anyMatch(e -> e.equalsIgnoreCase(etiqueta)))
                .sorted(Comparator.comparing(Noticia::getFechaPublicacion).reversed())
                .map(NoticiaResponse::desde)
                .toList();
    }

    public NoticiaResponse obtener(Long id) {
        return NoticiaResponse.desde(buscarEntidad(id));
    }

    public Noticia buscarEntidad(Long id) {
        return noticiaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Noticia no encontrada"));
    }

    @Transactional
    public NoticiaResponse crear(NoticiaRequest request, Usuario admin) {
        Noticia noticia = Noticia.builder()
                .titulo(request.getTitulo())
                .resumen(request.getResumen())
                .contenido(request.getContenido())
                .imagenUrl(request.getImagenUrl())
                .etiquetas(request.getEtiquetas())
                .publicadoPor(admin)
                .build();
        return NoticiaResponse.desde(noticiaRepository.save(noticia));
    }

    @Transactional
    public NoticiaResponse editar(Long id, NoticiaRequest request) {
        Noticia noticia = buscarEntidad(id);
        noticia.setTitulo(request.getTitulo());
        noticia.setResumen(request.getResumen());
        noticia.setContenido(request.getContenido());
        noticia.setImagenUrl(request.getImagenUrl());
        if (request.getEtiquetas() != null) noticia.setEtiquetas(request.getEtiquetas());
        return NoticiaResponse.desde(noticiaRepository.save(noticia));
    }

    @Transactional
    public void eliminar(Long id) {
        if (!noticiaRepository.existsById(id)) {
            throw new ResourceNotFoundException("Noticia no encontrada");
        }
        noticiaRepository.deleteById(id);
    }
}

File: frontend/lib/core/services/noticia_service.dart
import '../api/api_client.dart';
import '../models/noticia.dart';

class NoticiaService {
  final ApiClient _client;
  NoticiaService(this._client);

  Future<List<Noticia>> listar({String? etiqueta}) async {
    final data = await _client.get('/noticias', query: {'etiqueta': etiqueta});
    return (data as List).map((e) => Noticia.fromJson(e)).toList();
  }

  Future<Noticia> obtener(int id) async {
    final data = await _client.get('/noticias/$id');
    return Noticia.fromJson(data);
  }

  Future<Noticia> crear(Map<String, dynamic> body) async {
    final data = await _client.post('/admin/noticias', body: body);
    return Noticia.fromJson(data);
  }

  Future<Noticia> editar(int id, Map<String, dynamic> body) async {
    final data = await _client.put('/admin/noticias/$id', body: body);
    return Noticia.fromJson(data);
  }

  Future<void> eliminar(int id) => _client.delete('/admin/noticias/$id');
}
